src/main/java/org/apache/sysds/runtime/privacy/PrivacyPropagator.java

```diff
@@ -414,4 +591,25 @@ else if ( inst instanceof SqlCPInstruction )
 			instructionOutputNames = new String[]{((SqlCPInstruction) inst).getOutputVariableName()};
 		return instructionOutputNames;
 	}
+
+	private static ArrayList<CPOperand> getOutputOperands(Instruction inst){
+		// The order of the following statements is important
+		if ( inst instanceof MultiReturnParameterizedBuiltinCPInstruction )
+			return ((MultiReturnParameterizedBuiltinCPInstruction) inst).getOutputs();
+		else if ( inst instanceof MultiReturnBuiltinCPInstruction )
+			return ((MultiReturnBuiltinCPInstruction) inst).getOutputs();
+		else if ( inst instanceof ComputationCPInstruction )
+			return getSingletonList(((ComputationCPInstruction) inst).getOutput());
+		else if ( inst instanceof VariableCPInstruction )
+			return getSingletonList(((VariableCPInstruction) inst).getOutput());
+		else if ( inst instanceof SqlCPInstruction )
+			return getSingletonList(((SqlCPInstruction) inst).getOutput());
+		return new ArrayList<CPOperand>();
```

Comment:
not 100% sure, but the if else is instance of is probably slow, and there should be some alternative.

Reply:
The alternative is to have a method in Instruction called "getOutputs" and then implement it in the subclasses. This is however not the general design we use at the moment, but it would make much of the code cleaner to look at. For instance, most of the switch statements in this class could also be avoided if we changed the design and used Java inheritance instead.